assets/js/interactivity/store.js

```diff
@@ -32,12 +32,20 @@ const getSerializedState = () => {
 	return {};
 };
 
+const storeCallbacks = {};
 const rawState = getSerializedState();
 export const rawStore = { state: deepSignal( rawState ) };
 
 if ( typeof window !== 'undefined' ) window.store = rawStore;
 
-export const store = ( { state, ...block } ) => {
+export const store = ( { state, ...block }, callbacks = {} ) => {
 	deepMerge( rawStore, block );
 	deepMerge( rawState, state );
+	Object.entries( callbacks ).forEach( ( [ key, cb ] ) => {
+		( storeCallbacks[ key ] = storeCallbacks[ key ] || [] ).push( cb );
+	} );
+};
+
+export const runStoreCallbacks = ( key ) => {
+	storeCallbacks[ key ]?.forEach( ( cb ) => cb( rawStore ) );
```

Comment:
Here I've been considering how this function should work during client-side navigation, taking in mind that, in the future, some blocks will be loaded when navigating to new pages. To prevent calling the same callbacks again, maybe we can remove all registered callbacks once executed, and call `runStoreCallbacks()` again after each new render.

If this is not suitable for certain callback types (e.g., `afterNavigation`? 🤷 ), we could add a parameter to indicate that the executed callbacks should be removed afterward.

Reply:
Good point. Maybe we can use a `Map` to avoid adding the same reference twice, and mark the executed callbacks to avoid executing them again.

```js
const callbacks = {
  afterLoad: new Map(),
};

export const store = ({ state, ...block } = {}, { afterLoad } = {}) => {
  deepMerge(rawStore, block);
  deepMerge(rawState, state);
  if (afterLoad && !callbacks.afterLoad.has(afterLoad))
    callbacks.afterLoad.set(afterLoad, false);
};

export const runStoreCallbacks = (key) => {
  callbacks[key]?.entries(([cb, executed]) => {
    if (!executed) {
      cb(rawStore);
      callbacks[key].set(cb, true);
    }
  });
};
```

But let's wait to do so until we are actually testing this with full-page client-side navigation.